services/docker/docker-size.service.js

```diff
@@ -0,0 +1,134 @@
+'use strict'
+
+const Joi = require('@hapi/joi')
+const prettyBytes = require('pretty-bytes')
+const { nonNegativeInteger } = require('../validators')
+const { latest } = require('../version')
+const {
+  buildDockerUrl,
+  getDockerHubUser,
+  getMultiPageData,
+} = require('./docker-helpers')
+const { NotFound } = require('..')
+const { BaseJsonService } = require('..')
+
+const buildSchema = Joi.object({
+  name: Joi.string().required(),
+  full_size: nonNegativeInteger.required(),
+}).required()
+
+const pagedSchema = Joi.object({
+  count: nonNegativeInteger.required(),
+  results: Joi.array().items(
+    Joi.object({
+      name: Joi.string().required(),
+      full_size: nonNegativeInteger.required(),
+    })
+  ),
+}).required()
+
+const queryParamSchema = Joi.object({
+  sort: Joi.string()
+    .valid('date', 'semver')
+    .default('date'),
+}).required()
+
+module.exports = class DockerSize extends BaseJsonService {
+  static get category() {
+    return 'size'
+  }
+
+  static get route() {
+    return { ...buildDockerUrl('image-size', true), queryParamSchema }
+  }
+
+  static get examples() {
+    return [
+      {
+        title: 'Docker Image Size (latest by date)',
+        pattern: ':user/:repo',
+        namedParams: { user: 'fedora', repo: 'apache' },
+        staticPreview: this.render({ size: 126000000, sort: 'date' }),
+      },
+      {
+        title: 'Docker Image Size (latest SemVer)',
+        pattern: ':user/:repo',
+        namedParams: { user: 'fedora', repo: 'apache' },
+        queryParams: { sort: 'semver' },
+        staticPreview: this.render({ size: 136000000 }),
+      },
+      {
+        title: 'Docker Image Size (tag)',
+        pattern: ':user/:repo/:tag',
+        namedParams: { user: 'fedora', repo: 'apache', tag: 'latest' },
+        staticPreview: this.render({ size: 103000000 }),
+      },
+    ]
+  }
+
+  static get defaultBadgeData() {
+    return {
+      label: 'image size',
+      color: 'blue',
+    }
+  }
+
+  static render({ size }) {
+    return { message: prettyBytes(size) }
+  }
+
+  async fetch({ user, repo, tag, page }) {
+    page = page ? `&page=${page}` : ''
+    return this._requestJson({
+      schema: tag ? buildSchema : pagedSchema,
+      url: `https://registry.hub.docker.com/v2/repositories/${getDockerHubUser(
+        user
+      )}/${repo}/tags${
+        tag ? `/${tag}` : '?page_size=100&ordering=last_updated'
+      }${page}`,
+      errorMessages: { 404: 'image or tag not found' },
+    })
+  }
+
+  transform({ tag, sort, data }) {
+    if (!tag && sort === 'date') {
+      if (data.count === 0) {
```

Comment:
Don't we need these types of checks regardless of the tag/sort values? I believe in the case of `!tag`, a `data.count` of `0` would always be a `repository not found` scenario, but if data.count is 0 when the user specifies a tag, then it's a `repository or tag not found` since it could either be a bad user/repo name _or_ a bad tag name

Reply:
No we don't need them for tag/sort values, only some. I can see why the above routes were resulting in an error, I'll fix that.

In the case when a user specifies a tag for the size badge a different URI is hit (https://hub.docker.com/v2/repositories/shieldsio/shields/tags/broken), and results in a 404, this will return the `image or tag not found` error.

Reply:
>  and results in a 404

Ah okay, perfect!